File: dsa_leetcode_beginner/0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)
        left_product, right_product = 1, 1
        result = nums[0]

        for i in range(n):
            # if any of left or right product become 0, update it to 1
            if left_product == 0: 
                left_product = 1
            if right_product == 0: 
                right_product = 1

            # prefix product
            left_product *= nums[i]

            # suffix product
            right_product *= nums[n - 1 - i]

            result = max(ans, max(left_product, right_product))

        return result
        
        
#Question: https://leetcode.com/problems/maximum-product-subarray
#Blog: https://blog.unwiredlearning.com/maximum-product-subarray

File: dsa_leetcode_beginner/0152-maximum-product-subarray/0152-maximum-product-subarray.java

public class Solution {
    public int maxProduct(int[] nums) {
        int n = nums.length;
        int leftProduct = 1, rightProduct = 1;
        int result = nums[0];

        for (int i = 0; i < n; i++) {
            // If any of left or right product becomes 0, update it to 1
            if (leftProduct == 0) {
                leftProduct = 1;
            }
            if (rightProduct == 0) {
                rightProduct = 1;
            }

            // Prefix product
            leftProduct *= nums[i];

            // Suffix product
            rightProduct *= nums[n - 1 - i];

            // Update the result with the maximum value found
            result = Math.max(result, Math.max(leftProduct, rightProduct));
        }

        return result;
    }
}
